java-checks/src/main/java/org/sonar/java/checks/security/DebugFeatureEnabledCheck.java

```diff
@@ -19,32 +19,74 @@
  */
 package org.sonar.java.checks.security;
 
-import java.util.Collections;
+import java.util.Arrays;
+import java.util.Deque;
+import java.util.LinkedList;
 import java.util.List;
 import java.util.Objects;
 import javax.annotation.CheckForNull;
 import org.sonar.check.Rule;
 import org.sonar.java.checks.helpers.ExpressionsHelper;
 import org.sonar.plugins.java.api.IssuableSubscriptionVisitor;
+import org.sonar.plugins.java.api.semantic.Symbol;
 import org.sonar.plugins.java.api.tree.AnnotationTree;
 import org.sonar.plugins.java.api.tree.AssignmentExpressionTree;
+import org.sonar.plugins.java.api.tree.ClassTree;
 import org.sonar.plugins.java.api.tree.ExpressionTree;
 import org.sonar.plugins.java.api.tree.IdentifierTree;
+import org.sonar.plugins.java.api.tree.MemberSelectExpressionTree;
+import org.sonar.plugins.java.api.tree.MethodInvocationTree;
 import org.sonar.plugins.java.api.tree.Tree;
 
 @Rule(key = "S4507")
 public class DebugFeatureEnabledCheck extends IssuableSubscriptionVisitor {
 
   private static final String MESSAGE = "Make sure this debug feature is deactivated before delivering the code in production.";
 
+  private final Deque<Symbol.TypeSymbol> enclosingClass = new LinkedList<>();
+
   @Override
   public List<Tree.Kind> nodesToVisit() {
-    return Collections.singletonList(Tree.Kind.ANNOTATION);
+    return Arrays.asList(Tree.Kind.ANNOTATION, Tree.Kind.CLASS, Tree.Kind.METHOD_INVOCATION);
   }
 
   @Override
   public void visitNode(Tree tree) {
-    AnnotationTree annotation = (AnnotationTree) tree;
+    switch (tree.kind()) {
+      case ANNOTATION:
+        checkAnnotation((AnnotationTree) tree);
+        break;
+      case METHOD_INVOCATION:
+        checkMethodInvocation((MethodInvocationTree) tree);
+        break;
+      default:
+        ClassTree classTree = (ClassTree) tree;
+        enclosingClass.push(classTree.symbol());
+        break;
+    }
+  }
+
+  @Override
+  public void leaveNode(Tree tree) {
+    if (tree instanceof ClassTree) {
+      enclosingClass.pop();
+    }
+  }
+
+  private void checkMethodInvocation(MethodInvocationTree tree) {
+    if (tree.methodSelect().is(Tree.Kind.MEMBER_SELECT)) {
+      MemberSelectExpressionTree memberSelectExpressionTree = (MemberSelectExpressionTree) tree.methodSelect();
+      IdentifierTree identifierTree = memberSelectExpressionTree.identifier();
+      if (!enclosingClassExtendsThrowable()
+        && "printStackTrace".equals(identifierTree.name())
+        && calledOnTypeInheritedFromThrowable(memberSelectExpressionTree.expression())
+        && tree.arguments().isEmpty()) {
+        reportIssue(identifierTree, MESSAGE);
+      }
+    }
+  }
```

Comment:
:scream: the previous implementation of the rule was complex and not using the MethodMatchers API.
we should use a matcher like:
```
  private static final MethodMatchers PRINT_STACK_TRACE_MATCHER = MethodMatchers.create()
    .ofSubTypes("java.lang.Throwable").names("printStackTrace").addWithoutParametersMatcher().build();
```
them `checkMethodInvocation` will become far more easier to read:
```
  private void checkMethodInvocation(MethodInvocationTree tree) {
    if (PRINT_STACK_TRACE_MATCHER.matches(tree)) {
      reportIssue(ExpressionUtils.methodName(tree), MESSAGE);
    }
  }
```